onnx/defs/training/defs.cc

```diff
@@ -0,0 +1,171 @@
+Y with respect to W (Z). The user can compute gradient by inserting Gradient
+operator to form
+
+```
+W --> Conv --> H --> Gemm --> Y
+|      ^              ^
+|      |              |
+|      X              Z
+|      |              |
+|      |   .----------'
+|      |   |  (X/W/Z is the 1st/2nd/3rd input of Gradient as shown in "xs")
+|      v   v
+'---> Gradient(xs=["X", "W", "Z"], y=["Y"]) ---> dY/dX (1st output of Gradient)
+       |   |
+       |   '-----------------------------------> dY/dW (2nd output of Gradient)
+       |
+       '---------------------------------------> dY/dZ (3rd output of Gradient)
+```
+
+, where the content inside the braces of Gradient are attributes. The attribute
+"xs" specifies the necessary inputs to compute the variable specified by
+attribute "y". To compute gradient, the runtime should identify a sub-graph with
+inputs specified by "xs" and outputs containing "y"'s content and do a backward
+pass (or other auto-differentiation techniques whenever needed).
+
+Since X is not trainable tensor, the user can avoid the creation of dY/dX by
+assigning an empty string to the 1st output name of that Gradient. In other
+words, all Gradient's outputs are optional. Note that the concept of optional
+outputs can also be found in ONNX's RNN, GRU, and LSTM.
+
+Gradient operator can compute derivative against intermediate tensors. For
+example, the gradient of Y with respect to H can be done in
+
+```
+W --> Conv --> H --> Gemm --> Y
+       ^       |      ^
+       |       |      |
+       X       |      Z
+       .-------'      |
+       |   .----------'
+       |   | (H/Z is the 1st/2nd input of Gradient as shown in "xs")
+       v   v
+      Gradient(xs=["H", "Z"], y=["Y"])
+       |   |
+       |   '-----------------------------------> dY/dH (1st output of Gradient)
+       |
+       '---------------------------------------> dY/dZ (2nd output of Gradient)
+```
+
+It is possible to represent high-order differentiation using Gradient operator.
+An example for linear model is
+
+```
+W --> Gemm --> Y --> Loss --> O
+       ^              ^
+       |              |
+       X              L
+```
+
+To compute the 2nd order derivative of O with respect to W (denoted by
+d^2O/dW^2), one can do
+
+```
+W --> Gemm --> Y --> Loss --> O
+|      ^              ^
+|      |              |
+|      X              L
+|      |
+|      |
++------+--> Gradient(xs=["X", "W"], y=["Y"]) ---> dO/dX (1st output of Gradient)
+|      |      |
+|      |      '---> dO/dW (2nd output of Gradient)
+|      v
+'---> Gradient(xs=["X", "W"], y=["dO/dW"]) ---> d(dO/dW)dX (1st output of
+       |                                          Gradient)
+       |
+       |
+       '---> d^2O/dW^2 (2nd output of Gradient)
+```
+
+As mentioned above, the attributes "xs" and "y" are used to identify a graph,
+and we can feed different tensors to the identified graph. For,
+example, one can compute the gradient of Y with respect to H by substituting
```

Comment:
Are there typos here? I am a bit confused, because I am not sure what is Y_1, H_1, and the picture below shows W_1 and Z_1. Are you trying to say that it is not necessary for the inputs to be identical to the list xs specified in the attributes?

Reply:
If the above is the case, it would be helpful to add that "The implementation can be optimized in the common case where the inputs are identical to the attributes." … because if the inputs are not identical to the attributes, the implementation has to do extra work to create a duplicate of the forward-graph for the actual-inputs.

Reply:
Do you think we need to allow
```
 if the inputs are not identical to the attributes, the implementation has to do extra work to create a duplicate of the forward-graph for the actual-inputs
```
? Looks like it's not necessary.

Reply:
If we don't support it, that's okay with me (since I can't immediately think of use-cases where we would want to use it). But, then, the documentation should make this explicit (and implementations should check that the input-list and attribute-list match).